app/src/main/java/com/eis/phoneringer/structure/AppManager.java

```diff
@@ -0,0 +1,98 @@
+package com.eis.phoneringer.structure;
+
+import android.content.Context;
+import android.media.Ringtone;
+import android.os.Build;
+import android.os.Handler;
+import android.os.Looper;
+import android.widget.Toast;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.RequiresApi;
+
+import com.eis.smslibrary.SMSHandler;
+import com.eis.smslibrary.exceptions.InvalidSMSMessageException;
+import com.eis.smslibrary.exceptions.InvalidTelephoneNumberException;
+
+/**
+ * This is a singleton class used to manage a received RingCommand or to send one
+ *
+ * @author Alberto Ursino, Luca Crema, Alessandra Tonin, Marco Mariotto
+ */
+public class AppManager {
+
+    private static final String WRONG_PASSWORD_TOAST = "Wrong password";
+    private static final int TIMEOUT_TIME = 30 * 1000; //30 seconds
+
+    /**
+     * Instance of the class that is instantiated in getInstance method
+     */
+    private static AppManager instance = null;
+
+    /**
+     * Private constructor
+     */
+    private AppManager() {
+    }
+
+    /**
+     * @return the AppManager instance
+     */
+    public static AppManager getInstance() {
+        if (instance == null)
+            instance = new AppManager();
+        return instance;
+    }
+
+    /**
+     * If the password of the RingCommand received is valid then play ringtone for fixed amount of time
+     *
+     * @param context     of the application
+     * @param ringCommand received
+     * @param ringtone    to be played
+     */
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+    public void onRingCommandReceived(Context context, @NonNull RingCommand ringCommand, final Ringtone ringtone) {
+        final RingtoneHandler ringtoneHandler = RingtoneHandler.getInstance();
+
+        if (checkPassword(context, ringCommand)) {
+            ringtoneHandler.playRingtone(ringtone);
+            //Timer: the ringtone is playing for TIMEOUT_TIME seconds.
+            new Handler(Looper.getMainLooper()).postDelayed(new Runnable() {
+                @Override
+                public void run() {
+                    ringtoneHandler.stopRingtone(ringtone);
+                }
+            }, TIMEOUT_TIME);
+        } else {
+            Toast.makeText(context, WRONG_PASSWORD_TOAST, Toast.LENGTH_SHORT).show();
```

Comment:
Beautiful solution changing from a toast to an exception. I would make just a little edit, from `WrongPasswordException` to `InvalidPasswordException`. Same meaning, different way to say it, which in my opinion is more inline with the standard java api (think about `IllegalArgumentException`)

Reply:
I think this is a better solution than toast, but I also agree with @Cogno-IDU  about the exception name.

Reply:
I don't think `IllegalPasswordException` is the name we're looking for, this code checks if the password is correct, not if it's valid / with the right format. Anyway the name is still kinda intuitive and both can be used, but I'd still prefer `Wrong`

ps. I don't like at all the use of the adversative "but" in @tonin-alessandra comment: you are linking two positive sentences with a "but"! That's not correct! Should have used an "and"

Reply:
After receiving another revision, perhaps from @giovannivelludo, I think of fixing these expressions.

Reply:
@CremaLuca I also chose `Illegal` as a keyword because of the specifications
> Exception thrown when the password received is not valid

Not valid, as in, Invalid

In the end it comes down to personal preferences but I'll just make it consistent

Reply:
Agreed on `Ìllegal`

Reply:
Well, now you are in 3 that you want `Illigal` instead of `Wrong` so i'll fix it now.